### PR TITLE
Table view export

### PR DESCRIPTION
**Pull request by antares1470**
### Description of the Change

Have fixed the issue where copying and exporting data from the table view only was applied to the current page.

### Why Should This Be In Core?

Bug fix in Core

### Benefits

Users can copy and export the entire table instead of just the page they are currently on.

### Possible Drawbacks

Getting the selection variants working properly meant a cost in performance for export to excel

### Applicable Issues

#862

### CODE REVIEW

File: CoreTableView/src/au/gov/asd/tac/constellation/views/tableview2/TableViewPane.java [hunk 1]
```diff
@@ -1239,44 +1234,99 @@ public void updateSelection(final Graph graph, final TableViewState state) {
...
     }
+       
+    /**
+     * A version of the updateSelection(Graph, TableViewState) function which is
+     * to be run on the JavaFX Application Thread
+     * 
+     * @param graph the graph to read selection from.
+     * @param state the current table view state.
+     */
+    private void updateSelectionFromJAT(final Graph graph, final TableViewState state) {
```
Comment (arcturus2) [inline CoreTableView/src/au/gov/asd/tac/constellation/views/tableview2/TableViewPane.java:1263]: A real minor one but how about calling it `updateSelectionFromFxThread`?

File: CoreTableView/src/au/gov/asd/tac/constellation/views/tableview2/TableViewPane.java [hunk 2]
```diff
@@ -1239,44 +1234,99 @@ public void updateSelection(final Graph graph, final TableViewState state) {
...
+        synchronized (LOCK) {
+            if (graph != null && state != null) {
+
+                if (!Platform.isFxApplicationThread()) {
```
Comment (arcturus2) [inline CoreTableView/src/au/gov/asd/tac/constellation/views/tableview2/TableViewPane.java:1267]: So if it is meant to be run on the FX thread, how is this bit passing?

Comment (antares1470) [reply CoreTableView/src/au/gov/asd/tac/constellation/views/tableview2/TableViewPane.java:1267]: `Platform.isFxApplicationThread()` is a boolean function that returns true if we are running in the fx application thread. So the condition here returns true if we are NOT running on the fx application thread

File: CoreTableView/src/au/gov/asd/tac/constellation/views/tableview2/TableViewPane.java (hunk 1 above)
Comment (antares1470) [reply CoreTableView/src/au/gov/asd/tac/constellation/views/tableview2/TableViewPane.java:1263]: done

File: CoreTableView/src/au/gov/asd/tac/constellation/views/tableview2/TableViewPane.java (hunk 2 above)
Comment (arcturus2) [reply CoreTableView/src/au/gov/asd/tac/constellation/views/tableview2/TableViewPane.java:1267]: Doh! Thanks, I totally misread it.

**Review by arcturus2 — CHANGES_REQUESTED** [review]
Hi @antares1470 I got an `IllegalStateException` trying to export a graph with just 6 transactions to csv. I got a similar error trying to export to Excel.

```
java.lang.IllegalStateException: Not processing on the JavaFX Application Thread
	at au.gov.asd.tac.constellation.views.tableview2.TableViewPane.updateSelectionFromFXThread(TableViewPane.java:1268)
	at au.gov.asd.tac.constellation.views.tableview2.TableViewPane.createPage(TableViewPane.java:1044)
	at au.gov.asd.tac.constellation.views.tableview2.TableViewPane.lambda$paginate$62(TableViewPane.java:1054)
	at au.gov.asd.tac.constellation.views.tableview2.TableViewUtilities.getTableData(TableViewUtilities.java:117)
	at au.gov.asd.tac.constellation.views.tableview2.TableViewUtilities$ExportToCsvFilePlugin.execute(TableViewUtilities.java:258)
	at au.gov.asd.tac.constellation.plugins.templates.SimplePlugin.run(SimplePlugin.java:68)
[catch] at au.gov.asd.tac.constellation.graph.node.plugins.DefaultPluginEnvironment.lambda$executePluginLater$0(DefaultPluginEnvironment.java:129)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)
```

**Comment by arcturus2** [comment]
FYI the first time I started Constellation, created a sphere graph and then loaded the Table View it got stuck and looked like this:

image

I wasn't able to reproduce it though.

**Review by arcturus2 — APPROVED** [review]

**Comment by antares1470** [comment]
> 
> 
> FYI the first time I started Constellation, created a sphere graph and then loaded the Table View it got stuck and looked like this:
> 
> image
> 
> I wasn't able to reproduce it though.

Thanks @arcturus2 for testing. I did a little digging into this and when I tested, I too encountered the issue of table view not loading correctly (table view loaded but the data was either missing or in the wrong columns). Closing and loading table view a few times produced differing results before eventually loading correctly. 

I tried replicating in master and got no success so I went back to my branch to confirm whether it was my branch and got no success replicating the first time. I did get it to replicate the next time I ran it though. 

This makes me think that the issue you've identified may be something in my code so I'll do some more digging and see if I can resolve this.

**Comment by antares1470** [comment]
> Thanks @arcturus2 for testing. I did a little digging into this and when I tested, I too encountered the issue of table view not loading correctly (table view loaded but the data was either missing or in the wrong columns). Closing and loading table view a few times produced differing results before eventually loading correctly.
> 
> I tried replicating in master and got no success so I went back to my branch to confirm whether it was my branch and got no success replicating the first time. I did get it to replicate the next time I ran it though.
> 
> This makes me think that the issue you've identified may be something in my code so I'll do some more digging and see if I can resolve this.

Alright I think I have found the source of the problem and resolved it. My new update selection function had a synchronized despite the fact it would only be running on the fx thread (something I missed having initially copy pasted the original update selection function as a starting point for the new one).

**Pull request merged**